### PR TITLE
[SPARK-44714] Ease restriction of LCA resolution regarding queries with having

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?
This PR eases some restriction of LCA resolution regarding queries with having.

Previously LCA won't rewrite (to the new plan shape) when the whole queries contains `UnresolvedHaving`, in case it breaks the plan shape of `UnresolvedHaving - Aggregate` that can be recognized by other rules. But this limitation is too strict and it causes some deadlock in having - lca - window queries. See https://issues.apache.org/jira/browse/SPARK-42936 for more details and examples.

With this PR, it will only skip LCA resolution on the `Aggregate` whose direct parent is `UnresolvedHaving`. This is enabled by a new bottom-up resolution without using the transform or resolve utility function.

This PR also recognizes a vulnerability related to `TEMP_RESOVLED_COLUMN` and comments in the code. It should be considered as future work.

### Why are the changes needed?
More complete functionality and better user experience.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 1]
```diff
@@ -131,138 +131,166 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
+      // UnresolvedHaving, to avoid breaking the shape pattern  `UnresolvedHaving - Aggregate`
+      // matched by ResolveAggregateFunctions. See SPARK-42936 for more details.
+      case u @ UnresolvedHaving(_, agg: Aggregate)
+        if agg.aggregateExpressions.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+        u.copy(child = agg.mapChildren(apply0))
 
-        case agg @ Aggregate(groupingExpressions, aggregateExpressions, _)
-          if ruleApplicableOnOperator(agg, aggregateExpressions)
-            && aggregateExpressions.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+      case p @ Project(projectList, child) if ruleApplicableOnOperator(p, projectList)
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:147]: Completely copied the original code except adding the final line to recursively iterating children

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 2]
```diff
@@ -131,138 +131,166 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
-          if (!aggregateExpressions.forall(eligibleToLiftUp)) {
-            return agg
+          p.copy(
+            projectList = outerProjectList.toSeq,
+            child = Project(innerProjectList.toSeq, child)
+          )
+        }
+        newPlan.mapChildren(apply0)
+
+      case agg @ Aggregate(groupingExpressions, aggregateExpressions, _)
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:196]: Completely copied the original code except adding the final line to recursively applying on the new Project

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 3]
```diff
@@ -131,95 +131,97 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
+      // matched by ResolveAggregateFunctions. See SPARK-42936 and SPARK-44714 for more details.
+      case u @ UnresolvedHaving(_, agg: Aggregate) =>
+        u.copy(child = agg.mapChildren(apply0))
 
-        case agg @ Aggregate(groupingExpressions, aggregateExpressions, _)
-          if ruleApplicableOnOperator(agg, aggregateExpressions)
-            && aggregateExpressions.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+      case pOriginal: Project if ruleApplicableOnOperator(pOriginal, pOriginal.projectList)
+          && pOriginal.projectList.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+        val p @ Project(projectList, child) = pOriginal.mapChildren(apply0)
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:148]: Bottom-up resolution. The rest of code is fully copied and has no change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 4]
```diff
@@ -131,95 +131,97 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
+            projectList = outerProjectList.toSeq,
+            child = Project(innerProjectList.toSeq, child)
+          )
+        }
+
+      case aggOriginal: Aggregate
+        if ruleApplicableOnOperator(aggOriginal, aggOriginal.aggregateExpressions)
+          && aggOriginal.aggregateExpressions.exists(
+            _.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+        val agg @ Aggregate(groupingExpressions, aggregateExpressions, _) =
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:199]: Bottom up resolution. The rest of code is fully copied and has no change (except the one I commented out).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 5]
```diff
@@ -131,95 +131,97 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
+              && !groupingExpressions.exists(_.semanticEquals(s)) => false
+            // Manually skip detection on function itself because it can be an aggregate function.
+            // This is to avoid expressions like sum(salary) over () eligible to lift up.
+            case WindowExpression(function, spec) =>
+              function.children.forall(eligibleToLiftUp) && eligibleToLiftUp(spec)
+            case e => e.children.forall(eligibleToLiftUp)
+          }
+        }
+        if (!aggregateExpressions.forall(eligibleToLiftUp)) {
+          agg
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:223]: Previously this line was `return agg`. That could be risky to `return` in a closure.. But for `apply0` with or without return should be the same.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**